cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -33,42 +33,85 @@
 #include "arrow/util/logging.h"
 #include "arrow/util/string.h"
 
-namespace arrow {
-namespace fs {
+namespace arrow::fs {
+
+namespace Blobs = Azure::Storage::Blobs;
+namespace Core = Azure::Core;
+namespace DataLake = Azure::Storage::Files::DataLake;
+namespace Storage = Azure::Storage;
 
 // -----------------------------------------------------------------------
 // AzureOptions Implementation
 
 AzureOptions::AzureOptions() = default;
 
+AzureOptions::~AzureOptions() = default;
+
 bool AzureOptions::Equals(const AzureOptions& other) const {
-  return (account_dfs_url == other.account_dfs_url &&
-          account_blob_url == other.account_blob_url &&
-          credentials_kind == other.credentials_kind &&
-          default_metadata == other.default_metadata);
+  // TODO(GH-38598): update here when more auth methods are added.
+  const bool ret = backend == other.backend &&
+                   default_metadata == other.default_metadata &&
+                   account_blob_url_ == other.account_blob_url_ &&
+                   account_dfs_url_ == other.account_dfs_url_ &&
+                   credential_kind_ == other.credential_kind_;
+  if (!ret) {
+    return false;
+  }
+  switch (credential_kind_) {
+    case CredentialKind::kAnonymous:
+      return true;
+    case CredentialKind::kStorageSharedKeyCredential:
+      return storage_shared_key_credential_->AccountName ==
+             other.storage_shared_key_credential_->AccountName;
+  }
+  DCHECK(false);
+  return false;
```

Comment:
How about using `default`?

```suggestion
    default:
      DCHECK(false);
      return false;
  }
```

We may want to add a comment something like "never happen".

Reply:
Not having the `default` means the compiler will complain when a new enum entry is added reminding the programmer that they need to add the case to this loop. Putting the `DCHECK(false); return false;` after the switch is required by MSVC (if warning are being treated as errors).

Reply:
Or perhaps use `arrow/util/unreachable.h`?

Reply:
I ended up with this:

```cpp
Result<std::unique_ptr<Blobs::BlobServiceClient>> AzureOptions::MakeBlobServiceClient()
    const {
  switch (credential_kind_) {
    case CredentialKind::kAnonymous:
      break;
    case CredentialKind::kStorageSharedKeyCredential:
      return std::make_unique<Blobs::BlobServiceClient>(account_blob_url_,
                                                        storage_shared_key_credential_);
  }
  return Status::Invalid("AzureOptions doesn't contain a valid auth configuration");
}
```

Reply:
+1, thank you.